File: src/language_profile.rs
use std::collections::BTreeSet;
use std::error::Error;
use std::fmt;

use crate::{LinkId, LinkMetadata, LinkNetwork, LinkType, ParseConfiguration, TranslationRuleSet};

const PROFILE_TERM: &str = "language-profile";
const PROFILE_LINK_TYPE_TERM: &str = "language-profile:link-type";
const PROFILE_CONCEPT_TERM: &str = "language-profile:concept";
const PROFILE_TRANSLATION_RULE_TERM: &str = "language-profile:translation-rule";
const PROFILE_DIAGNOSTIC_TERM: &str = "language-profile:unsupported-feature";

/// Per-language capability profile for restricting transforms to supported features.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct LanguageProfile {
    name: String,
    language: String,
    link_types: BTreeSet<LinkType>,
    concepts: BTreeSet<String>,
    translation_rules: BTreeSet<String>,
}

impl LanguageProfile {
    /// Creates an empty profile for a target language.
    #[must_use]
    pub fn new(name: impl Into<String>, language: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            language: language.into(),
            link_types: BTreeSet::new(),
            concepts: BTreeSet::new(),
            translation_rules: BTreeSet::new(),
        }
    }

    /// Built-in JavaScript same-language profile.
    #[must_use]
    pub fn javascript() -> Self {
        let mut profile = Self::new("JavaScript", "JavaScript");
        for link_type in [
            LinkType::Link,
            LinkType::Reference,
            LinkType::Relation,
            LinkType::Language,
            LinkType::Grammar,
            LinkType::Type,
            LinkType::Concept,
            LinkType::Syntax,
            LinkType::Field,
            LinkType::Trivia,
            LinkType::Token,
            LinkType::Document,
            LinkType::Semantic,
            LinkType::Region,
            LinkType::Object,
        ] {
            profile = profile.with_link_type(link_type);
        }
        profile
    }

    /// Looks up a built-in profile by name.
    #[must_use]
    pub fn builtin(name: &str) -> Option<Self> {
        match name.to_ascii_lowercase().as_str() {
            "javascript" | "js" => Some(Self::javascript()),
            _ => None,
        }
    }

    /// Computes a profile domain from a translation rule set.
    ///
    /// Rule query link-type filters become supported link types, query term
    /// filters become supported concept/feature terms, and every rule name is
    /// recorded as a supported translation rule.
    #[must_use]
    pub fn from_rule_set(
        name: impl Into<String>,
        language: impl Into<String>,
        rule_set: &TranslationRuleSet,
    ) -> Self {
        let mut profile = Self::new(name, language);
        for rule in rule_set.rules() {
            profile = profile.with_translation_rule(rule.name());
            if let Some(link_type) = rule.query().link_type_filter() {
                profile = profile.with_link_type(link_type);
            }
            if let Some(term) = rule.query().term_filter() {
                profile = profile.with_concept(term);
            }
        }
        profile
    }

    /// Profile name.
    #[must_use]
    pub fn name(&self) -> &str {
        &self.name
    }

    /// Target language this profile constrains.
    #[must_use]
    pub fn language(&self) -> &str {
        &self.language
    }

    /// Supported link types.
    #[must_use]
    pub const fn link_types(&self) -> &BTreeSet<LinkType> {
        &self.link_types
    }

    /// Supported concept or feature terms.
    #[must_use]
    pub const fn concepts(&self) -> &BTreeSet<String> {
        &self.concepts
    }

    /// Supported translation rule names.
    #[must_use]
    pub const fn translation_rules(&self) -> &BTreeSet<String> {
        &self.translation_rules
    }

    /// Returns a copy with a supported link type.
    #[must_use]
    pub fn with_link_type(mut self, link_type: LinkType) -> Self {
        self.link_types.insert(link_type);
        self
    }

    /// Returns a copy with a supported concept or feature term.
    #[must_use]
    pub fn with_concept(mut self, concept: impl Into<String>) -> Self {
        self.concepts.insert(concept.into());
        self
    }

    /// Returns a copy with a supported translation rule name.
    #[must_use]
    pub fn with_translation_rule(mut self, rule: impl Into<String>) -> Self {
        self.translation_rules.insert(rule.into());
        self
    }

    /// Whether this profile supports a link type.
    #[must_use]
    pub fn supports_link_type(&self, link_type: LinkType) -> bool {
        self.link_types.contains(&link_type)
    }

    /// Whether this profile supports a concept or feature term.
    #[must_use]
    pub fn supports_concept(&self, concept: &str) -> bool {
        self.concepts.contains(concept)
    }

    /// Whether this profile supports a translation rule name.
    #[must_use]
    pub fn supports_translation_rule(&self, rule: &str) -> bool {
        self.translation_rules.contains(rule)
    }

    /// Declares this profile as queryable links inside a network.
    pub fn declare_in(&self, network: &mut LinkNetwork) -> LanguageProfileLinks {
        let profile = self.profile_link(network).unwrap_or_else(|| {
            network.insert_link(
                [],
                LinkMetadata::new()
                    .with_link_type(LinkType::Semantic)
                    .with_named(true)
                    .with_term(PROFILE_TERM)
                    .with_language(&self.language)
                    .with_definition(&self.name),
            )
        });
        let mut capabilities = Vec::new();

        for link_type in &self.link_types {
            capabilities.push(self.ensure_capability_link(
                network,
                profile,
                PROFILE_LINK_TYPE_TERM,
                &link_type.to_string(),
            ));
        }
        for concept in &self.concepts {
            capabilities.push(self.ensure_capability_link(
                network,
                profile,
                PROFILE_CONCEPT_TERM,
                concept,
            ));
        }
        for rule in &self.translation_rules {
            capabilities.push(self.ensure_capability_link(
                network,
                profile,
                PROFILE_TRANSLATION_RULE_TERM,
                rule,
            ));
        }

        LanguageProfileLinks {
            profile,
            capabilities,
        }
    }

    /// Validates that all typed links in a network stay inside this profile.
    ///
    /// # Errors
    ///
    /// Returns [`LanguageProfileViolation`] for the first unsupported link
    /// type found in identifier order.
    pub fn validate_network(&self, network: &LinkNetwork) -> Result<(), LanguageProfileViolation> {
        for link in network.links() {
            if let Some(link_type) = link.metadata().link_type() {
                if !self.supports_link_type(link_type) {
                    return Err(LanguageProfileViolation::new(
                        format!("link type `{link_type}`"),
                        format!(
                            "Profile `{}` for `{}` does not support link type `{link_type}`.",
                            self.name, self.language
                        ),
                    ));
                }
            }

            if self.concepts.is_empty()
                || !matches!(
                    link.metadata().link_type(),
                    Some(LinkType::Concept | LinkType::Semantic)
                )
            {
                continue;
            }
            let Some(term) = link.metadata().term() else {
                continue;
            };
            if is_profile_control_term(term) || self.supports_concept(term) {
                continue;
            }
            return Err(LanguageProfileViolation::new(
                format!("concept `{term}`"),
                format!(
                    "Profile `{}` for `{}` does not support concept `{term}`.",
                    self.name, self.language
                ),
            ));
        }
        Ok(())
    }

    pub(crate) fn validate_transform_result(
        &self,
        network: &LinkNetwork,
    ) -> Result<(), LanguageProfileViolation> {
        self.validate_network(network)?;

        let source = network.reconstruct_text();
        if source.is_empty() {
            return Ok(());
        }

        let parsed = LinkNetwork::parse(&source, &self.language, ParseConfiguration::default());
        let report = parsed.verify_full_match(None);
        if report.issues().is_empty() {
            Ok(())
        } else {
            Err(LanguageProfileViolation::new(
                format!("{} syntax", self.language),
                format!(
                    "Profile `{}` for `{}` rejects source text that is not valid {}.",
                    self.name, self.language, self.language
                ),
            ))
        }
    }

    pub(crate) fn insert_diagnostic(
        &self,
        network: &mut LinkNetwork,
        violation: &LanguageProfileViolation,
        subject: Option<LinkId>,
    ) -> LinkId {
        let profile = self.declare_in(network).profile();
        let metadata = LinkMetadata::new()
            .with_link_type(LinkType::Semantic)
            .with_named(true)
            .with_term(PROFILE_DIAGNOSTIC_TERM)
            .with_language(&self.language)
            .with_definition(violation.to_string());

        match subject {
            Some(subject) => network.insert_link([profile, subject], metadata),
            None => network.insert_link([profile], metadata),
        }
    }

    fn profile_link(&self, network: &LinkNetwork) -> Option<LinkId> {
        network
            .links()
            .find(|link| {
                link.metadata().link_type() == Some(LinkType::Semantic)
                    && link.metadata().term() == Some(PROFILE_TERM)
                    && link.metadata().language() == Some(self.language())
                    && link.metadata().definition() == Some(self.name())
            })
            .map(crate::Link::id)
    }

    fn ensure_capability_link(
        &self,
        network: &mut LinkNetwork,
        profile: LinkId,
        term: &str,
        definition: &str,
    ) -> LinkId {
        if let Some(existing) = network
            .links()
            .find(|link| {
                link.references() == [profile]
                    && link.metadata().link_type() == Some(LinkType::Semantic)
                    && link.metadata().term() == Some(term)
                    && link.metadata().language() == Some(self.language())
                    && link.metadata().definition() == Some(definition)
            })
            .map(crate::Link::id)
        {
            return existing;
        }

        network.insert_link(
            [profile],
            LinkMetadata::new()
                .with_link_type(LinkType::Semantic)
                .with_named(true)
                .with_term(term)
                .with_language(&self.language)
                .with_definition(definition),
        )
    }
}

/// Links inserted when a language profile is declared in a network.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct LanguageProfileLinks {
    profile: LinkId,
    capabilities: Vec<LinkId>,
}

impl LanguageProfileLinks {
    /// Root profile link.
    #[must_use]
    pub const fn profile(&self) -> LinkId {
        self.profile
    }

    /// Capability child links.
    #[must_use]
    pub fn capabilities(&self) -> &[LinkId] {
        &self.capabilities
    }
}

/// A profile validation failure that can be recorded as a diagnostic link.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct LanguageProfileViolation {
    feature: String,
    message: String,
}

impl LanguageProfileViolation {
    fn new(feature: impl Into<String>, message: impl Into<String>) -> Self {
        Self {
            feature: feature.into(),
            message: message.into(),
        }
    }

    /// Unsupported feature that caused the violation.
    #[must_use]
    pub fn feature(&self) -> &str {
        &self.feature
    }
}

impl fmt::Display for LanguageProfileViolation {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            formatter,
            "{} Unsupported feature: {}.",
            self.message, self.feature
        )
    }
}

impl Error for LanguageProfileViolation {}

fn is_profile_control_term(term: &str) -> bool {
    term.starts_with("language-profile")
        || term.starts_with("translation-rule:")
        || term == "translation-rule"
        || term == "translation-rule-set"
}

File: js/src/language-profile.js
import { LinkMetadata, LinkType } from './primitives.js';

const PROFILE_TERM = 'language-profile';
const PROFILE_LINK_TYPE_TERM = 'language-profile:link-type';
const PROFILE_CONCEPT_TERM = 'language-profile:concept';
const PROFILE_TRANSLATION_RULE_TERM = 'language-profile:translation-rule';

// Rust `LinkType` carries a richer enum than the JS primitives currently expose.
// We map the JS members that have a Rust counterpart and emit the Rust
// `Display` spelling (lowercased) for capability definitions and diagnostics so
// declared/validated networks stay byte-compatible with the Rust port.
const LINK_TYPE_DISPLAY = Object.freeze({
  [LinkType.Concept]: 'concept',
  [LinkType.Dynamic]: 'dynamic',
  [LinkType.Field]: 'field',
  [LinkType.Language]: 'language',
  [LinkType.Object]: 'object',
  [LinkType.Relation]: 'relation',
  [LinkType.Semantic]: 'semantic',
  [LinkType.SourceToken]: 'token',
  [LinkType.Syntax]: 'syntax',
  [LinkType.Trivia]: 'trivia',
});

function linkTypeDisplay(linkType) {
  return LINK_TYPE_DISPLAY[linkType] ?? String(linkType).toLowerCase();
}

// Stable ordering key for a link type, mirroring Rust's `BTreeSet<LinkType>`
// iteration by emitting a deterministic (here: display-name) order.
function linkTypeKey(linkType) {
  return linkTypeDisplay(linkType);
}

function sortedStrings(values) {
  return [...values].sort((left, right) => (left < right ? -1 : left > right ? 1 : 0));
}

function sortedLinkTypes(values) {
  return [...values].sort((left, right) => {
    const leftKey = linkTypeKey(left);
    const rightKey = linkTypeKey(right);
    return leftKey < rightKey ? -1 : leftKey > rightKey ? 1 : 0;
  });
}

// Profile and capability links are identified by their `definition`, mirroring
// the Rust profile's `with_definition`/`definition()`. The JS `LinkMetadata`
// (primitives.js) carries a first-class `definition` field for full parity.
function buildCapabilityMetadata(term, language, definition) {
  return LinkMetadata.new()
    .withLinkType(LinkType.Semantic)
    .withNamed(true)
    .withTerm(term)
    .withLanguage(language)
    .withDefinition(definition);
}

function metadataDefinition(metadata) {
  return metadata.definition;
}

function sameReferences(references, expected) {
  if (references.length !== expected.length) {
    return false;
  }
  return references.every((reference, index) => reference.equals(expected[index]));
}

function isProfileControlTerm(term) {
  return (
    term.startsWith('language-profile') ||
    term.startsWith('translation-rule:') ||
    term === 'translation-rule' ||
    term === 'translation-rule-set'
  );
}

/// Per-language capability profile for restricting transforms to supported features.
export class LanguageProfile {
  constructor(name, language) {
    this._name = String(name);
    this._language = String(language);
    this._linkTypes = new Set();
    this._concepts = new Set();
    this._translationRules = new Set();
    this._fallbacks = new Map();
  }

  /// Creates an empty profile for a target language.
  static new(name, language) {
    return new LanguageProfile(name, language);
  }

  /// Built-in JavaScript same-language profile.
  static javascript() {
    let profile = new LanguageProfile('JavaScript', 'JavaScript');
    for (const linkType of [
      LinkType.Relation,
      LinkType.Language,
      LinkType.Concept,
      LinkType.Syntax,
      LinkType.Field,
      LinkType.Trivia,
      LinkType.SourceToken,
      LinkType.Semantic,
      LinkType.Object,
    ]) {
      profile = profile.withLinkType(linkType);
    }
    return profile;
  }

  /// Looks up a built-in profile by name.
  static builtin(name) {
    switch (String(name).toLowerCase()) {
      case 'javascript':
      case 'js':
        return LanguageProfile.javascript();
      default:
        return undefined;
    }
  }

  /// Computes a profile domain from a translation rule set.
  static fromRuleSet(name, language, ruleSet) {
    let profile = new LanguageProfile(name, language);
    for (const rule of ruleSet.rules) {
      profile = profile.withTranslationRule(rule.name);
      const query = rule.query;
      if (query && query.linkType !== undefined) {
        profile = profile.withLinkType(query.linkType);
      }
      if (query && query.term !== undefined) {
        profile = profile.withConcept(query.term);
      }
    }
    return profile;
  }

  static from_rule_set(name, language, ruleSet) {
    return LanguageProfile.fromRuleSet(name, language, ruleSet);
  }

  _clone() {
    const copy = new LanguageProfile(this._name, this._language);
    copy._linkTypes = new Set(this._linkTypes);
    copy._concepts = new Set(this._concepts);
    copy._translationRules = new Set(this._translationRules);
    copy._fallbacks = new Map(this._fallbacks);
    return copy;
  }

  /// Profile name.
  name() {
    return this._name;
  }

  /// Target language this profile constrains.
  language() {
    return this._language;
  }

  /// Supported link types (deterministic order, mirroring Rust `BTreeSet`).
  linkTypes() {
    return sortedLinkTypes(this._linkTypes);
  }

  link_types() {
    return this.linkTypes();
  }

  /// Supported concept or feature terms (sorted).
  concepts() {
    return sortedStrings(this._concepts);
  }

  /// Supported translation rule names (sorted).
  translationRules() {
    return sortedStrings(this._translationRules);
  }

  translation_rules() {
    return this.translationRules();
  }

  /// Unsupported concepts mapped to their documented lossy fallback (sorted by key).
  fallbacks() {
    return new Map(sortedStrings(this._fallbacks.keys()).map((key) => [key, this._fallbacks.get(key)]));
  }

  /// Returns a copy with a supported link type.
  withLinkType(linkType) {
    const copy = this._clone();
    copy._linkTypes.add(linkType);
    return copy;
  }

  with_link_type(linkType) {
    return this.withLinkType(linkType);
  }

  /// Returns a copy with a supported concept or feature term.
  withConcept(concept) {
    const copy = this._clone();
    copy._concepts.add(String(concept));
    return copy;
  }

  with_concept(concept) {
    return this.withConcept(concept);
  }

  /// Returns a copy with a supported translation rule name.
  withTranslationRule(rule) {
    const copy = this._clone();
    copy._translationRules.add(String(rule));
    return copy;
  }

  with_translation_rule(rule) {
    return this.withTranslationRule(rule);
  }

  /// Returns a copy that records an unsupported concept and its lossy fallback.
  withConceptFallback(concept, fallback) {
    const copy = this._clone();
    copy._fallbacks.set(String(concept), String(fallback));
    return copy;
  }

  with_concept_fallback(concept, fallback) {
    return this.withConceptFallback(concept, fallback);
  }

  /// The documented lossy fallback for a concept, or undefined.
  conceptFallback(concept) {
    return this._fallbacks.get(concept);
  }

  concept_fallback(concept) {
    return this.conceptFallback(concept);
  }

  /// Whether this profile supports a link type.
  supportsLinkType(linkType) {
    return this._linkTypes.has(linkType);
  }

  supports_link_type(linkType) {
    return this.supportsLinkType(linkType);
  }

  /// Whether this profile supports a concept or feature term.
  supportsConcept(concept) {
    return this._concepts.has(concept);
  }

  supports_concept(concept) {
    return this.supportsConcept(concept);
  }

  /// Whether this profile supports a translation rule name.
  supportsTranslationRule(rule) {
    return this._translationRules.has(rule);
  }

  supports_translation_rule(rule) {
    return this.supportsTranslationRule(rule);
  }

  /// Declares this profile as queryable links inside a network.
  declareIn(network) {
    const profile =
      this._profileLink(network) ??
      network.insertLink([], buildCapabilityMetadata(PROFILE_TERM, this._language, this._name));

    const capabilities = [];

    for (const linkType of this.linkTypes()) {
      capabilities.push(
        this._ensureCapabilityLink(network, profile, PROFILE_LINK_TYPE_TERM, linkTypeDisplay(linkType)),
      );
    }
    for (const concept of this.concepts()) {
      capabilities.push(this._ensureCapabilityLink(network, profile, PROFILE_CONCEPT_TERM, concept));
    }
    for (const rule of this.translationRules()) {
      capabilities.push(
        this._ensureCapabilityLink(network, profile, PROFILE_TRANSLATION_RULE_TERM, rule),
      );
    }

    return new LanguageProfileLinks(profile, capabilities);
  }

  declare_in(network) {
    return this.declareIn(network);
  }

  /// Validates that all typed links in a network stay inside this profile.
  validateNetwork(network) {
    for (const link of network.links()) {
      const metadata = link.metadata();
      const linkType = metadata.linkType;
      if (linkType !== undefined && !this.supportsLinkType(linkType)) {
        throw new LanguageProfileViolation(
          `link type \`${linkTypeDisplay(linkType)}\``,
          `Profile \`${this._name}\` for \`${this._language}\` does not support link type \`${linkTypeDisplay(linkType)}\`.`,
        );
      }

      if (
        this._concepts.size === 0 ||
        !(linkType === LinkType.Concept || linkType === LinkType.Semantic)
      ) {
        continue;
      }
      const term = metadata.term;
      if (term === undefined) {
        continue;
      }
      if (isProfileControlTerm(term) || this.supportsConcept(term)) {
        continue;
      }
      throw new LanguageProfileViolation(
        `concept \`${term}\``,
        `Profile \`${this._name}\` for \`${this._language}\` does not support concept \`${term}\`.`,
      );
    }
  }

  validate_network(network) {
    return this.validateNetwork(network);
  }

  _profileLink(network) {
    const found = network.links().find((link) => {
      const metadata = link.metadata();
      return (
        metadata.linkType === LinkType.Semantic &&
        metadata.term === PROFILE_TERM &&
        metadata.language === this._language &&
        metadataDefinition(metadata) === this._name
      );
    });
    return found?.id();
  }

  _ensureCapabilityLink(network, profile, term, definition) {
    const existing = network.links().find((link) => {
      const metadata = link.metadata();
      return (
        sameReferences(link.references(), [profile]) &&
        metadata.linkType === LinkType.Semantic &&
        metadata.term === term &&
        metadata.language === this._language &&
        metadataDefinition(metadata) === definition
      );
    });
    if (existing) {
      return existing.id();
    }

    return network.insertLink(
      [profile],
      buildCapabilityMetadata(term, this._language, definition),
    );
  }
}

/// Links inserted when a language profile is declared in a network.
export class LanguageProfileLinks {
  constructor(profile, capabilities = []) {
    this._profile = profile;
    this._capabilities = capabilities;
  }

  /// Root profile link.
  profile() {
    return this._profile;
  }

  /// Capability child links.
  capabilities() {
    return [...this._capabilities];
  }
}

/// A profile validation failure that can be recorded as a diagnostic link.
export class LanguageProfileViolation extends Error {
  constructor(feature, message) {
    super(`${message} Unsupported feature: ${feature}.`);
    this.name = 'LanguageProfileViolation';
    this._feature = feature;
    this._message = message;
  }

  /// Unsupported feature that caused the violation.
  feature() {
    return this._feature;
  }

  toString() {
    return this.message;
  }
}
